text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1042 @@
+    * All of the above have happened in some form, so I believe it's time to have the T-lang meeting again.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+    * We've already handled this since the last RFC, there are no more implementation concerns.
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+    * This RFC's syntax is entirely unrelated to discussions on effect syntax.
+        * If we get an effect system, it may be desirable to allow expressing const traits with the effect syntax, this design is forward compatible with that.
+        * If we get an effect system, we will still want this shorthand, just like we allow you to write:
+            * `T: Iterator<Item = U>` and don't require `where T: Iterator, <T as Iterator>::Item = U`.
+            * `T: Iterator<Item: Debug>` and don't require `where T: Iterator, <T as Iterator>::Item: Debug`.
+    * RTN for per-method bounds: `T: Trait<some_fn(..): [const] Fn(A, B) -> C>` could supplement this feature in the future.
+        * Alternatively `where <T as Trait>::some_fn(..): [const]` or `where <T as Trait>::some_fn \ {const}`.
+        * Very verbose (need to specify arguments and return type).
+        * Want short hand sugar anyway to make it trivial to change a normal function to a const function by just adding some minor annotations.
+        * Significantly would delay const trait stabilization (by years).
+        * Usually requires editing the trait anyway, so there's no "can constify impls without trait author opt in" silver bullet.
+    * New RTN-like per-method bounds: `T: Trait<some_fn(_): [const]>`.
+        * Unclear if soundly possible.
+        * Unclear if possible without incurring significant performance issues for all code (may need tracking new information for all functions out there).
+        * Still requires editing traits.
+        * Still want the `[const] Trait` sugar anyway.
+
+## Should we start out without `const Trait` bounds
+
+We do not need to immediately allow using methods on generic parameters of const fn, as a lot of const code is nongeneric.
+
+The following example could be made to work with just const traits and const trait impls.
+
+```rust
+struct MyStruct(i32);
+
+impl PartialEq for MyStruct {
+    const fn eq(&self, other: &MyStruct) -> bool {
+        self.0 == other.0
+    }
+}
+
+const fn foo() {
+    let a = MyStruct(1);
+    let b = MyStruct(2);
+    if a == b {}
+}
+```
+
+Things like `Option::map` or `PartialEq` for arrays/tuples could not be made const without const trait bounds,
+as they need to actually call the generic `FnOnce` argument or nested `PartialEq` impls.
+
+
+# Future possibilities
```

Comment:
Since this RFC proposes the `Destruct` trait, it might be appropriate to mention allowing `?Destruct` without `const`[^1] as a future possibility.

[^1]: for example, opting out of `Destruct` in favor of a future `async Destruct`, or having a type require some other data to be dropped